keps/sig-network/1880-service-subnet/README.md

```diff
@@ -0,0 +1,558 @@
+
+```
+<<[UNRESOLVED allocation mode]>>
+
+* and allocated object can have different status: allocated, reserved, free, ...
+
+* on service creation, if the NodePort or ClusterIP field is set:
+1. check that the object doesn't exist in the apiserver and Create a new one (owned by the Service)
+2. check if the object exists and the Status is Free and change the status.
+
+* on service creation, if the NodePort or ClusterIP field is not set, allocate a random one from the defined range in the corresponding Object (IPRange or NodePortRange)
+
+There can be different model to allocate a random resource:
+
+1. Same as today, the apiserver can keep an index with current status and start with a random offset to
+find a free resources.
+
+2. A controller create a pool of free resource Objects and the apiserver picks one of them, this adds 
+more complexity to the model, but will improve the time to allocate a new free resources since is just
+popping an object from a list.
+
+<<[/UNRESOLVED]>>
+```
+
+## Design Details
+
+
+### Cluster IPs
+
+#### IP Ranges
+
+Define two new objects, a new IPRange Object that defines the range used to allocate Services Cluster IPs
+
+```go
+// IPRange defines a range of IPs using CIDR format (192.168.0.0/24 or 2001:db2::0/64).
+type IPRange struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   IPRangeSpec   `json:"spec,omitempty"`
+	Status IPRangeStatus `json:"status,omitempty"`
+}
+
+// IPRangeSpec describe how the IPRange's specification looks like.
+type IPRangeSpec struct {
+	// Range of IPs in CIDR format (192.168.0.0/24 or 2001:db2::0/64).
+	Range string `json:"range"`
+	// Primary indicates if this is the primary allocator to be used by the
+	// apiserver to allocate IP addresses.
+	// NOTE this can simplify the Service strategy logic so we don't have to infer
+	// the primary allocator, it also may allow to switch between primary families in
+	// a cluster, but this looks like a loooong shot.
+	// +optional
+	Primary bool `json:"primary,omitempty"`
+}
+
+// IPRangeStatus defines the observed state of IPRange.
+type IPRangeStatus struct {
+	// Free represent the number of IP addresses that are not allocated in the Range.
+	// +optional
+	Free int64 `json:"free,omitempty"`
+}
+```
+
+Reserve well-known-names for the IPRanges objects, that represent current allocators service-subnet-ranges:
+
+```go
+// default allocator well-known-names
+const (
+	DefaultIPv4Allocator = "allocator.ipv4.k8s.io"
+	DefaultIPv6Allocator = "allocator.ipv6.k8s.io"
+)
+```
+
+On the controlplane controller, add a new function similar to the existing ones used for [bootraping the default kubernetes services and namespaces](https://github.com/kubernetes/kubernetes/blob/ff34bd246b26bef622d35de68f3c143055e3aac7/pkg/controlplane/controller.go#L246-L251), to create the IPRange from the vale defined in the corresponding flag with the service-iprange-subnet
```

Comment:
Do we *always* sync flags into objects or only if no objects exist?

I.e., what happens if:

* boot apiserver with flags
* auto-create range objects from flags
* user changes the range objects
* apiserver restarts (still has flags) 

?  Hopefully the "IfNeeded" in the code below means "if there are no ranges for this family" ?

This would add to the idea that a single range could hold 2 families.  The conversion from flags to raanges would be atomic.